### PR TITLE
🎨  🏗 Move from JPA repository to JDBCTemplate and split out detail code into its own Controller/Server/Repository

### PR DESCRIPTION
**Pull request by nickmcmahon01**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt [hunk 1]
```diff
@@ -23,28 +22,28 @@ data class ShiftOvertimeDto @JsonCreator constructor(
...
         companion object {
 
-                fun from(it: ShiftOvertime): ShiftOvertimeDto {
-                        return ShiftOvertimeDto(
+                fun from(it: Detail): DetailDto {
+                        return DetailDto(
                                 it.date,
                                 it.detailStartTimeInSeconds,
                                 it.detailEndTimeInSeconds,
                                 ShiftType.OVERTIME.name,
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt:40]: If this is now a generic detail dto, we shoud hardcode the shift type to overtime elsewhere

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt [hunk 2]
```diff
@@ -7,14 +7,14 @@ import io.swagger.annotations.ApiModelProperty
 import org.slf4j.LoggerFactory
 import uk.gov.justice.digital.hmpps.csr.api.domain.ActionType
 import uk.gov.justice.digital.hmpps.csr.api.domain.ShiftType
-import uk.gov.justice.digital.hmpps.csr.api.model.ShiftDetail
+import uk.gov.justice.digital.hmpps.csr.api.model.DetailNotification
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt:10]: Is it worth dropping the term "notification" from `DetailNotification` and `ShiftNotification`? It doesn't feel like it adds any value and makes both a bit more wordy. Come to think of it, I imagine the concept of a "notification" only really comes into play in `cmd-api`, in this service we're just getting modifications.

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt [hunk 3]
```diff
@@ -7,14 +7,14 @@ import io.swagger.annotations.ApiModelProperty
...
 import uk.gov.justice.digital.hmpps.csr.api.model.ShiftNotification
 import uk.gov.justice.digital.hmpps.csr.api.service.NotificationService
 import java.time.LocalDate
 import java.time.LocalDateTime
 
 @ApiModel(description = "ShiftNotification")
-data class ShiftNotificationDto @JsonCreator constructor(
+data class NotificationDto @JsonCreator constructor(
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt:17]: `ShiftDto`?

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+data class Detail(
+
+        var date: LocalDate,
+
+        var detailStartTimeInSeconds: Long,
+
+        var detailEndTimeInSeconds: Long,
+
+        var activity: String
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt:13]: Does a detail need a type?

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/DetailNotification.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.digital.hmpps.csr.api.model
+
+import java.time.LocalDate
+import java.time.LocalDateTime
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Id
+
+data class DetailNotification(
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/DetailNotification.kt:9]: From reading the naming, I cannot figure out what the difference between a Detail and a DetailNotification is

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt (hunk 1 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt:40]: https://github.com/ministryofjustice/cmd-csr-procs/blob/master/Notification_Shift_Detail/GET_MODIFIED_DETAIL.sql#L60

See what mons comes up with but he can do something like that I think ^^^^

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt (hunk 2 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/NotificationDto.kt:10]: you're right. I think the endpoint shouldn't be /notifications even, its /planunit/{id}/shift?from=&to=&modifiedOnly=true and  /planunit/{id}/detail?from=&to=&modifiedOnly=true 
what do you think?

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import io.swagger.annotations.ApiModel
+import io.swagger.annotations.ApiModelProperty
+import uk.gov.justice.digital.hmpps.csr.api.domain.ActionType
+import uk.gov.justice.digital.hmpps.csr.api.domain.ShiftType
+import uk.gov.justice.digital.hmpps.csr.api.model.Detail
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+@ApiModel(description = "DetailDto")
+data class DetailDto @JsonCreator constructor(
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt:14]: Is this different than what is being spewed out by the c# app? Do we have a record of all the changes that need to be implemented to wire these up together?

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        val actionType: ActionType?
+) {
+        companion object {
+
+                fun from(it: Detail): DetailDto {
+                        return DetailDto(
+                                it.quantumId,
+                                it.shiftModified,
+                                it.shiftDate,
+                                it.shiftType?.let { type -> ShiftType.from(type) },
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/dto/DetailDto.kt:54]: ohh I really like this syntax

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package uk.gov.justice.digital.hmpps.csr.api.model
+
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+data class Detail(
+
+        val quantumId: String?,
+
+        val shiftModified: LocalDateTime?,
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt:10]: shiftModified or detailModified?

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/repository/DetailRepository.kt [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                GET_DETAILS,
+                MapSqlParameterSource()
+                        .addValue("from", from)
+                        .addValue("to", to)
+                        .addValue("quantumId", quantumId),
+                detailsRowMapper
+        )
+    }
+
+    fun getModifiedShifts(planUnit: String): Collection<Detail> {
```
Comment (astroash) [inline src/main/java/uk/gov/justice/digital/hmpps/csr/api/repository/DetailRepository.kt:25]: what's the difference between `getModifiedShifts` and `getModifiedDetails`? It sounds like the former shouldn't be on this repository

File: src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt (hunk 8 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/justice/digital/hmpps/csr/api/model/Detail.kt:10]: This value does actually seem to be shift modified.
https://github.com/ministryofjustice/cmd-csr-procs/blob/master/Notification_Shift_Detail/GET_MODIFIED_DETAIL.sql#L51
There is something in those comments that I think we need to talk to Mons about it, they're talking about including 'deleted' rows but in the code they're hardcoding all details to 'EDIT'

**Review by astroash — APPROVED** [review]

**Pull request merged**